src/SmartFormat/ZString/ZCharArray.cs

```diff
@@ -0,0 +1,332 @@
+﻿//
+// Copyright SmartFormat Project maintainers and contributors.
+// Licensed under the MIT license.
+//
+
+using System;
+using System.Buffers;
+
+namespace SmartFormat.ZString;
+
+/// <summary>
+/// A lightweight container that rents a char array as a buffer from an <see cref="ArrayPool{T}"/> and returns it when disposed.
+/// It simplifies passing around the buffer without intermediate memory allocations.
+/// <para/>
+/// Note that failing to dispose this struct after use will result in a memory leak.
+/// <para/>
+/// <see cref="ZCharArray"/> is thread-safe.
+/// </summary>
+public struct ZCharArray : IDisposable
+{
+    // ArrayPool is thread-safe
+    private static readonly ArrayPool<char>
+        Pool = ArrayPool<char>.Create(MaxBufferCapacity, 50);
+
+    private static readonly object Locker = new();
+
+    private char[]? _bufferArray;
+    private int _currentLength;
+
+    /// <summary>
+    /// The default capacity of the array.
+    /// </summary>
+    public const int DefaultBufferCapacity = 1_000_000;
+
+    /// <summary>
+    /// The maximum capacity of the array.
+    /// </summary>
+    public const int MaxBufferCapacity = 10_000_000;
+
+    /// <summary>
+    /// The maximum size of the stack-allocated buffer.
+    /// </summary>
+    public const int StackAllocCharBufferSizeLimit = 256;
+
+    /// <summary>
+    /// Creates a new <see cref="ZCharArray"/> with a length of <see cref="DefaultBufferCapacity"/>>.
+    /// </summary>
+    public ZCharArray() : this(DefaultBufferCapacity)
+    {
+    }
+
+    /// <summary>
+    /// Creates a new <see cref="ZCharArray"/> with the specified length.
+    /// </summary>
+    /// <param name="length">The length of the array.</param>
+    public ZCharArray(int length)
+    {
+        lock (Locker)
```

Comment:
Is it worth having a version without the lock to avoid the overhead when you know you won't be using multi-threading?

Reply:
Yes, I started with a generic version of the class, and left the locks in after changing it to `char` specific.
The overhead is small (11 ns), but locking for writing `char`s doesn't seem to be useful here. So I'll remove it.

```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3737/23H2/2023Update/SunValley3)
   13th Gen Intel Core i7-13700K, 1 CPU, 24 logical and 16 physical cores
   .NET SDK 8.0.302
     [Host]   : .NET 8.0.6 (8.0.624.26715), X64 RyuJIT AVX2
     .NET 8.0 : .NET 8.0.6 (8.0.624.26715), X64 RyuJIT AVX2
   
   Job=.NET 8.0  Runtime=.NET 8.0
   
   | Method   | Mean       | Error     | StdDev    | Allocated |
   |--------- |-----------:|----------:|----------:|----------:|
   | WithLock | 11.4621 ns | 0.0160 ns | 0.0149 ns |         - |
   | NoLock   |  0.0000 ns | 0.0000 ns | 0.0000 ns |         - |
```